### DISCUSSION TITLE
There is no "print PDF" and reviewer/approver action(for approve or reject)

### DISCUSSION THREAD

**Discussion by tobyWKW** [AOT-Technologies/forms-flow-ai#1403] [Q&A] [answered]
Why is there no reviewer/approver action(for approve or reject) after submitting a form from a client account to a reviewer/approver account?(I open submitted the form via a reviewer/approver account)...Is there anything I need to set up when designing the form/workflow? Thank you.

**Comment by auslin-aot** [comment]
Hi @tobyWKW , can you please refer to this answer **https://github.com/AOT-Technologies/forms-flow-ai/issues/237#issuecomment-1072023158** if Review Section not showing in the form.

In the workflow, the reviewer task should be assigned to the corresponding reviewer group.
![image](https://user-images.githubusercontent.com/99173163/233385056-483e6a93-932f-4915-aa63-1606ad67eb35.png)

**Reply by tobyWKW** [reply]
Thank you for the reply. I get this part now. But do I need to write other code in order for the "approve"/"reject" to actually do something? For example, when "Approve" is selected by the approver/reviewer, the workflow move "forward", and if "Reject" is selected the form is sent back to the client to refill the form and re-submit etc. Thank you.

**Reply by auslin-aot** [reply]
yes. you need to do in the form & workflow. 

In the form's review section capture reviewed/returned/rejected.
![image](https://user-images.githubusercontent.com/99173163/233900090-121cb416-30d7-4d01-9aa1-cc3f8748139d.png)

![image](https://user-images.githubusercontent.com/99173163/233900284-0b4b143d-1b29-4374-975b-bfd3a0978d7c.png)

In the workflow, have condition flows matching the reviewer action.
![image](https://user-images.githubusercontent.com/99173163/233899913-a5a2ca90-197f-4b0f-8f80-4825fe001185.png)

![image](https://user-images.githubusercontent.com/99173163/233899991-c5bfe4f8-8f8c-405c-a66f-93caed642eef.png)

Please refer to the below links for sample workflows & forms
https://github.com/AOT-Technologies/forms-flow-ai-examples/blob/main/bpm-examples/camunda/approval-process.zip  https://github.com/AOT-Technologies/forms-flow-ai-examples/tree/main/forms-examples

**Reply by tobyWKW** [reply]
Thank you again for the reply. I understand that part now. But how do I make the form return to the client/clerk if the approver/reviewer rejected it? And also how do I make the form proceed to the next stage in the workflow(let's say it is an update application task then to end event) if the approver/reviewer approved it? Do I need to write code for the Button Custom Logic? If so, how? Thanks a lot. I really appreciate the help.

**Reply by ansulalu-aot** [reply, marked as answer]
Hi @tobyWKW , let me explain you with a workflow.

![image](https://user-images.githubusercontent.com/116330596/234489365-6bb91b93-9585-430b-81b7-c667aaf89c33.png)

When a new form is submitted the start event of the workflow will be invoked.

Provide below listeners and script to the sequence flow to execute the business logic :

- BPMFormDataPipelineListener - https://github.com/AOT-Technologies/forms-flow-ai/blob/master/forms-flow-bpm/starter-examples/listeners/bpmformdatapipeline-readme.md

- ApplicationStateListener- https://github.com/AOT-Technologies/forms-flow-ai/blob/master/forms-flow-bpm/starter-examples/listeners/applicationstate-readme.md

- FormBPMFilteredDataPipelineListener -  https://github.com/AOT-Technologies/forms-flow-ai/blob/master/forms-flow-bpm/starter-examples/listeners/formbpmfiltereddatapipeline-readme.md

![image](https://user-images.githubusercontent.com/116330596/234489778-debcd278-9cdc-4881-8c10-660897d8648f.png)

The form will be routed to the user task labeled "Review Applications." Following the review process, it will either be marked as "Reviewed" or "Returned."

- In the event that the application is returned, it will be directed to the receive task labeled "Application Resubmitted" for resubmission of the application. 

- The conditional flow must have a value that corresponds to the reviewer's action of "Returned".

![image](https://user-images.githubusercontent.com/116330596/234496292-8de2bc98-7961-46fa-a005-1696e4a7dca0.png)

- In the receive task(Application Resubmitted), include both the Message and Execution listeners.

![image](https://user-images.githubusercontent.com/116330596/234496514-8fd4de46-95a0-4f59-87c9-e9eedff24a8b.png)

- If the application is reviewed, it will be directed to the user task labeled "Approve/Reject Applications."

- The conditional flow must have a value that corresponds to the reviewer's action of "Reviewed".

![image](https://user-images.githubusercontent.com/116330596/234497085-d82e5c63-8344-484f-9303-b59780e09bf7.png)

- Once the application reaches the user task labeled "Approve/Reject Applications," a gateway can be utilized to determine the appropriate course of action, such as approving, rejecting, or returning the application.

- The conditions in the outgoing sequence flows from the gateway must match the action taken by the reviewer, whether it be "Approved," "Rejected," or "Returned."

![image](https://user-images.githubusercontent.com/116330596/234495386-6e5bafd1-d8fb-49ce-9f0b-97b573de2923.png)
![image](https://user-images.githubusercontent.com/116330596/234495542-6e2a9d89-174e-4c14-ae5b-bc0279d68192.png)
![image](https://user-images.githubusercontent.com/116330596/234495721-3e720aee-c261-4b81-8e65-7773014135dd.png)

- If approved/ rejected the application status will be updated accordingly and will end the process.

![image](https://user-images.githubusercontent.com/116330596/234495913-98429f80-ea3b-49d0-bab6-ea483afcace8.png)

- You can refer the below link to know about listeners :
	https://aot-technologies.github.io/forms-flow-ai-doc/#camunda_listeners

- Please refer to the below link for sample workflows :
       https://github.com/AOT-Technologies/forms-flow-ai-examples/blob/main/bpm-examples/camunda/approval-process.zip 

**In the form, you should do the following :**

In the form’s review section capture reviewed/returned/rejected.

![image](https://user-images.githubusercontent.com/116330596/234492935-8ebfc80b-5b05-4020-83ce-620f351229bc.png)
![image](https://user-images.githubusercontent.com/116330596/234492689-c8f2d7de-2a63-41db-817b-84ea52ff72f4.png)

**Note**: Remember the value in the form and the action specified in the condition of the workflow be the same.

**In Button Custom Logic**, you should write the code to handle the submission of form data to a server, and emit events based on whether the submission was successful or not.

![image](https://user-images.githubusercontent.com/116330596/234495026-1f05e836-2aa1-406c-ad4a-03ae08477d59.png)

- Please refer to the below link for sample forms :
         https://github.com/AOT-Technologies/forms-flow-ai-examples/tree/main/forms-examples
